File: backend/src/main/java/com/interview/service/CppCompilerService.java
package com.interview.service;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

@Slf4j
@Service
public class CppCompilerService {

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ExecutionResult {
        private boolean compiled;
        private String compilerMessage;
        private String stdout;
        private String stderr;
        private int exitCode;
        private long timeMs;
        private boolean timeout;
    }

    /**
     * Compiles and runs C++ source code with the given input and returns the result.
     */
    public ExecutionResult compileAndRun(String sourceCode, String input) {
        String uuid = UUID.randomUUID().toString();
        boolean isWindows = System.getProperty("os.name").toLowerCase().contains("win");
        
        File workingDir = new File(System.getProperty("user.dir"), "temp_compile");
        if (!workingDir.exists()) {
            workingDir.mkdirs();
        }

        File sourceFile = new File(workingDir, "Solution_" + uuid + ".cpp");
        String exeName = "Solution_" + uuid + (isWindows ? ".exe" : "");
        File exeFile = new File(workingDir, exeName);

        try {
            // 1. Write source code to file
            Files.writeString(sourceFile.toPath(), sourceCode, StandardCharsets.UTF_8);

            // 2. Compile source code using g++
            ProcessBuilder compilePb = new ProcessBuilder("g++", "-O3", sourceFile.getName(), "-o", exeName);
            compilePb.directory(workingDir);
            
            Process compileProcess = compilePb.start();
            String compilerError = readStream(compileProcess.getErrorStream());
            String compilerOut = readStream(compileProcess.getInputStream());
            
            boolean compileCompleted = compileProcess.waitFor(10, TimeUnit.SECONDS);
            if (!compileCompleted) {
                compileProcess.destroyForcibly();
                return ExecutionResult.builder()
                        .compiled(false)
                        .compilerMessage("Compilation timed out (exceeded 10 seconds).")
                        .build();
            }

            int compileExitCode = compileProcess.exitValue();
            if (compileExitCode != 0) {
                return ExecutionResult.builder()
                        .compiled(false)
                        .compilerMessage(compilerError.isEmpty() ? compilerOut : compilerError)
                        .build();
            }

            // 3. Execute compiled binary
            ProcessBuilder runPb;
            if (isWindows) {
                runPb = new ProcessBuilder(exeFile.getAbsolutePath());
            } else {
                runPb = new ProcessBuilder("./" + exeName);
            }
            runPb.directory(workingDir);

            long startTime = System.currentTimeMillis();
            Process runProcess = runPb.start();

            // Inject standard input
            if (input != null && !input.isEmpty()) {
                try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(runProcess.getOutputStream(), StandardCharsets.UTF_8))) {
                    writer.write(input);
                    writer.flush();
                }
            }

            // Wait with a 2-second timeout
            boolean runCompleted = runProcess.waitFor(2, TimeUnit.SECONDS);
            long timeMs = System.currentTimeMillis() - startTime;

            if (!runCompleted) {
                runProcess.destroyForcibly();
                return ExecutionResult.builder()
                        .compiled(true)
                        .timeout(true)
                        .timeMs(timeMs)
                        .stderr("Time Limit Exceeded (exceeded 2.0 seconds)")
                        .stdout("")
                        .build();
            }

            String stdout = readStream(runProcess.getInputStream());
            String stderr = readStream(runProcess.getErrorStream());
            int runExitCode = runProcess.exitValue();

            return ExecutionResult.builder()
                    .compiled(true)
                    .timeout(false)
                    .stdout(stdout)
                    .stderr(stderr)
                    .exitCode(runExitCode)
                    .timeMs(timeMs)
                    .build();

        } catch (Exception e) {
            log.error("Error during compilation/execution step", e);
            return ExecutionResult.builder()
                    .compiled(false)
                    .compilerMessage("System execution error: " + e.getMessage())
                    .build();
        } finally {
            // 4. Cleanup temporary files
            try {
                if (sourceFile.exists()) {
                    sourceFile.delete();
                }
                if (exeFile.exists()) {
                    exeFile.delete();
                }
            } catch (Exception e) {
                log.warn("Failed to delete temp files: {}", e.getMessage());
            }
        }
    }

    private String readStream(InputStream stream) throws IOException {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            StringBuilder builder = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                builder.append(line).append("\n");
            }
            return builder.toString().trim();
        }
    }
}

File: backend/src/services/cpp-compiler.service.js
const { spawn, exec } = require('child_process');
const fs = require('fs').promises;
const path = require('path');
const os = require('os');
const crypto = require('crypto');

const compileAndRun = async (sourceCode, input) => {
  const uuid = crypto.randomUUID();
  const isWindows = os.platform() === 'win32';
  const workingDir = path.resolve('temp_compile');
  
  // Ensure the temporary compilation folder exists
  await fs.mkdir(workingDir, { recursive: true });
  
  const sourceFileName = `Solution_${uuid}.cpp`;
  const exeFileName = `Solution_${uuid}${isWindows ? '.exe' : ''}`;
  
  const sourceFilePath = path.join(workingDir, sourceFileName);
  const exeFilePath = path.join(workingDir, exeFileName);
  
  try {
    // 1. Write source code to disk
    await fs.writeFile(sourceFilePath, sourceCode, 'utf8');
    
    // 2. Compile source code using g++
    const compileCmd = `g++ -O3 ${sourceFileName} -o ${exeFileName}`;
    
    const compileResult = await new Promise((resolve) => {
      exec(compileCmd, { cwd: workingDir, timeout: 10000 }, (error, stdout, stderr) => {
        if (error) {
          resolve({
            compiled: false,
            compilerMessage: stderr.trim() || stdout.trim() || error.message
          });
        } else {
          resolve({ compiled: true });
        }
      });
    });
    
    if (!compileResult.compiled) {
      return {
        compiled: false,
        compilerMessage: compileResult.compilerMessage,
        stdout: '',
        stderr: '',
        exitCode: 1,
        timeMs: 0,
        timeout: false
      };
    }
    
    // 3. Execute the compiled binary
    const runBinaryPath = isWindows ? exeFilePath : `./${exeFileName}`;
    const startTime = Date.now();
    
    const runResult = await new Promise((resolve) => {
      const child = spawn(runBinaryPath, [], { cwd: workingDir });
      
      let stdoutData = '';
      let stderrData = '';
      let timer = null;
      let didTimeout = false;
      
      // Set execution timeout (2 seconds)
      timer = setTimeout(() => {
        didTimeout = true;
        child.kill('SIGKILL');
      }, 2000);
      
      // Pipe input stream
      if (input) {
        child.stdin.write(input);
        child.stdin.end();
      }
      
      child.stdout.on('data', (data) => {
        stdoutData += data.toString();
      });
      
      child.stderr.on('data', (data) => {
        stderrData += data.toString();
      });
      
      child.on('close', (code) => {
        clearTimeout(timer);
        const endTime = Date.now();
        const timeMs = endTime - startTime;
        
        if (didTimeout) {
          resolve({
            compiled: true,
            compilerMessage: '',
            stdout: '',
            stderr: 'Time Limit Exceeded (exceeded 2.0 seconds)',
            exitCode: -1,
            timeMs,
            timeout: true
          });
        } else {
          resolve({
            compiled: true,
            compilerMessage: '',
            stdout: stdoutData.trim(),
            stderr: stderrData.trim(),
            exitCode: code === null ? 0 : code,
            timeMs,
            timeout: false
          });
        }
      });
      
      child.on('error', (err) => {
        clearTimeout(timer);
        resolve({
          compiled: true,
          compilerMessage: `Runtime execution error: ${err.message}`,
          stdout: '',
          stderr: err.message,
          exitCode: -1,
          timeMs: Date.now() - startTime,
          timeout: false
        });
      });
    });
    
    return runResult;
    
  } catch (error) {
    console.error('Error during compilation/execution step:', error);
    return {
      compiled: false,
      compilerMessage: `System execution error: ${error.message}`,
      stdout: '',
      stderr: '',
      exitCode: -1,
      timeMs: 0,
      timeout: false
    };
  } finally {
    // 4. Cleanup temp files
    try {
      await fs.unlink(sourceFilePath).catch(() => {});
      await fs.unlink(exeFilePath).catch(() => {});
    } catch (cleanupError) {
      console.warn('Failed to delete temp compilation files:', cleanupError.message);
    }
  }
};

module.exports = {
  compileAndRun
};
